src/lib/visualization/cardinality_writer.cpp

```diff
@@ -0,0 +1,218 @@
+#include "cardinality_writer.hpp"
+
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "constant_mappings.hpp"
+#include "expression/expression_utils.hpp"
+#include "expression/lqp_subquery_expression.hpp"
+#include "expression/pqp_subquery_expression.hpp"
+#include "logical_query_plan/abstract_non_query_node.hpp"
+#include "logical_query_plan/join_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "operators/limit.hpp"
+#include "operators/projection.hpp"
+#include "operators/table_scan.hpp"
+
+namespace opossum {
+
+CardinalityWriter::CardinalityWriter() {
+  // We can guarantee the LQP never changes during visits and thus avoid redundant estimations for subplans
+  _cardinality_estimator.guarantee_bottom_up_construction();
+}
+
+void CardinalityWriter::write_cardinalities(const std::vector<std::shared_ptr<AbstractLQPNode>>& lqp_roots,
+                                            const std::vector<std::shared_ptr<AbstractOperator>>& pqp_roots,
+                                            const std::string& benchmark_item_id) {
+  // Traverse LQP and PQP trees and save every node along the way in _lqp_nodes and _pqp_nodes
+  _build_lqp_graph(lqp_roots);
+  _build_pqp_graph(pqp_roots);
+
+  const auto lqp_nodes_size = _lqp_nodes.size();
+  const auto pqp_nodes_size = _pqp_nodes.size();
+
+  // We assume that when we traverse LQP and PQP the same way, we will get the same nodes.
+  // All nodes that don't match up are ignored. This works well for the majority of tested LQPs and PQPs.
+  const auto number_nodes = std::min(_lqp_nodes.size(), _pqp_nodes.size());
+
+  // Open estimation CSV file
+  // APPEND mode is necessary, since the method will be called multiple times during benchmark execution
+  // and the previous results shouldn't be erased.
+  std::ofstream output_estimations;
+  output_estimations.open("./benchmark_cardinality_estimation.csv", std::ios_base::app);
+
+  // Open debug CSV file for PQP and LQP node number information, which can be used to identify possible
+  // benchmark items with mismatched nodes and for manual correction of the matching.
+  std::ofstream output_debug;
+  output_debug.open("./benchmark_cardinality_estimation_debug_info.csv", std::ios_base::app);
+
+  // If files are empty, write CSV headers.
+  // Use semicolon as seperator instead of comma to be able to work with database values containing commas
+  std::ifstream csv_file;
+  csv_file.open("./benchmark_cardinality_estimation.csv");
+  if (csv_file.peek() == std::ifstream::traits_type::eof()) {
```

Comment:
No, since we create a new `CardinalityWriter` every time we call the method (see `src/benchmarklib/benchmark_sql_executor.cpp` line 174). We think that with the current architecture of the BenchmarkRunner this would be quite cumbersome to improve, since it would have to be created and held by the topmost execution threat and passed along for each new benchmark item.